drivers/retained_mem/retained_mem_nrf_gpregret.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_gpregret
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_nrf_gpregret, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_MULTITHREADING
+struct nrf_gpregret_data {
+	struct k_mutex lock;
+};
+#endif
+
+struct nrf_gpregret_config {
+	uint8_t *addr;
+	size_t size;
+};
+
+static inline void nrf_gpregret_lock_take(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static inline void nrf_gpregret_lock_release(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_unlock(&data->lock);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static int nrf_gpregret_init(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+#endif
+
+	return 0;
+}
+
+static ssize_t nrf_gpregret_size(const struct device *dev)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+	return (ssize_t)config->size;
+}
+
+static int nrf_gpregret_read(const struct device *dev, size_t offset, uint8_t *buffer, size_t size)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+	nrf_gpregret_lock_take(dev);
+
+	memcpy(buffer, (config->addr + offset), size);
```

Comment:
`offset` (and size) needs to be checked to prevent gaining unrestricted access to RAM with the use of the API, from the user space.